plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3FileSystemExchangeStorage.java

```diff
@@ -0,0 +1,545 @@
+
+        private int currentPartNumber;
+        private CompletableFuture<PutObjectResponse> directUploadFuture;
+        private CompletableFuture<CreateMultipartUploadResponse> multiPartUploadIdFuture;
+        private final List<CompletableFuture<CompletedPart>> multiPartUploadFutures = new ArrayList<>();
+
+        public S3ExchangeStorageWriter(S3AsyncClient s3AsyncClient, String bucketName, String key, int partSize, Optional<SecretKey> secretKey)
+        {
+            this.s3AsyncClient = requireNonNull(s3AsyncClient, "s3AsyncClient is null");
+            this.bucketName = requireNonNull(bucketName, "bucketName is null");
+            this.key = requireNonNull(key, "key is null");
+            this.partSize = partSize;
+            this.secretKey = requireNonNull(secretKey, "secretKey is null");
+        }
+
+        @Override
+        public CompletableFuture<?> write(Slice slice)
+        {
+            // skip multipart upload if there would only be one part
+            if (slice.length() < partSize && multiPartUploadIdFuture == null) {
+                PutObjectRequest.Builder putObjectRequestBuilder = PutObjectRequest.builder()
+                        .bucket(bucketName)
+                        .key(key);
+                S3RequestUtil.configureEncryption(secretKey, putObjectRequestBuilder);
+                directUploadFuture = s3AsyncClient.putObject(putObjectRequestBuilder.build(), ByteBufferAsyncRequestBody.fromByteBuffer(slice.toByteBuffer()));
+                return directUploadFuture;
+            }
+
+            if (multiPartUploadIdFuture == null) {
+                multiPartUploadIdFuture = createMultipartUpload();
+            }
+
+            final int partNum = ++currentPartNumber;
+            CompletableFuture<CompletedPart> uploadFuture = multiPartUploadIdFuture
+                    .thenCompose(createMultipartUploadResponse -> uploadPart(createMultipartUploadResponse.uploadId(), slice, partNum));
+            multiPartUploadFutures.add(uploadFuture);
+
+            return uploadFuture;
+        }
+
+        @Override
+        public CompletableFuture<?> abort()
+        {
+            if (multiPartUploadIdFuture == null) {
+                if (directUploadFuture != null) {
+                    directUploadFuture.cancel(true);
+                }
+                return completedFuture(null);
+            }
+
+            multiPartUploadFutures.forEach(future -> future.cancel(true));
+            return abortMultipartUpload(getMultipartUploadId());
+        }
+
+        @Override
+        public CompletableFuture<?> finish()
+        {
+            if (multiPartUploadIdFuture == null) {
+                return requireNonNullElseGet(directUploadFuture, () -> completedFuture(null));
+            }
+
+            return allOf(multiPartUploadFutures.toArray(CompletableFuture[]::new))
+                    .thenApply(ignored -> multiPartUploadFutures.stream().map(CompletableFuture::join).collect(toImmutableList()))
```

Comment:
`getNow` requires an argument `valueIfAbsent`, which requires us to pass in a lambda instead of a function reference. Do you see if they make any difference here?

Reply:
Basically the assumption here is that the future is already finished. `join` would cause the future to wait if it's not finished. Thus it feels it is better encode the intention of getting a value that is expected to be present explicitly